builtin/game/falling.lua

```diff
@@ -532,13 +536,17 @@ function core.check_single_for_falling(p)
 				local success, _ = convert_to_falling_node(p, n)
 				return success
 			end
+			local d_falling = core.registered_nodes[n.name]
 			-- Otherwise only if the bottom node is considered "fall through"
 			if not same and
-					(not d_bottom.walkable or d_bottom.buildable_to) and
-					(core.get_item_group(n.name, "float") == 0 or
-					d_bottom.liquidtype == "none") then
-				local success, _ = convert_to_falling_node(p, n)
-				return success
+			    (not d_bottom.walkable or d_bottom.buildable_to) and
+			    ((core.get_item_group(n.name, "float") == 0 or
+			      d_bottom.liquidtype == "none") or
+				(core.get_item_group(n.name, "float") > 0 and
+				 d_falling.liquidtype == "source" and
+				 d_bottom.liquidtype ~= "source")) then
+			    local success, _ = convert_to_falling_node(p, n)
+			    return success
```

Comment:
What would you think about the following styling? I hope it does make the purpose of the conditions clearer by also documenting them:

```Lua
local do_float = core.get_item_group(n.name, "float") > 0
-- Otherwise only if the bottom node is considered "fall through"
if not same and
		(not d_bottom.walkable or d_bottom.buildable_to)
		and -- Take "float" group into consideration:
		(
			-- Fall through non-liquids
			not do_float or d_bottom.liquidtype == "none" or
			-- Only let sources fall through flowing liquids
			(do_float and d_falling.liquidtype == "source" and d_bottom.liquidtype ~= "source")
		) then

	local success, _ = convert_to_falling_node(p, n)
	return success
end
```

Reply:
It's my first time using this GitHub review feature to be honest.
Do you want me to apply your changes manually and add them as a separate commit?

Reply:
I took the liberty to change the code accordingly. Perhaps I should've used the code suggestion feature to simplify the integration with a single click.

Generally I would recommend you to keep a clean `master` branch that acts as a base for PR-specific branches. That way you will not have to restore the master branch (force push) after the PR is merged (or dismissed).